src/v/cluster/tm_stm.cc

```diff
@@ -72,16 +70,18 @@ tm_transaction_v1 downgrade_tx(const tm_transaction& tx) {
     return result;
 }
 
-model::record_batch
-serialize_tx(tm_transaction tx, use_new_version_tx_bool use_new_version_tx) {
+model::record_batch serialize_tx(
+  tm_transaction tx,
+  use_tx_version_with_last_pid_bool use_tx_version_with_last_pid) {
```

Comment:
Usually it's an anti-pattern to pass boolean flags to modify behavior of a method, just introduce two methods:
serialize_tx(tm_transaction) and serialize_tx(tm_transaction_v1) and use caller to dispatch. You can even get away with a single template version of serialize_tx.

Reply:
But I think for feature_* stuff it is right way to do different works.
What do you mean? As I understand in your case this logic will be in different place. Is it ok?

Reply:
Two places :) you also edit the callers of `serialize_tx`, let's simplify it:

```
template<class T>
static model::record_batch do_serialize_tx(T tx) {
    iobuf key;
    reflection::serialize(key, model::record_batch_type::tm_update);
    auto pid_id = tx.pid.id;
    auto tx_id = tx.id;
    reflection::serialize(key, pid_id, tx_id);

    iobuf value;
    reflection::serialize(value, T::version);
    reflection::serialize(value, std::move(tx));

    storage::record_batch_builder b(
      model::record_batch_type::tm_update, model::offset(0));
    b.add_raw_kv(std::move(key), std::move(value));
    return std::move(b).build();
}

static model::record_batch serialize_tx(tm_transaction tx) {
    if (use_new_tx_version()) {
        return do_serialize_tx(tx);
    }
    auto old_tx = downgrade_tx(tx);
    return do_serialize_tx(tx);
}
```